### ISSUE TITLE
Disable formatting via configuration file

### ISSUE THREAD

**Issue by qweasdzxc787** [BlankSpruce/gersemi#35]
Can we get this as an option in the configuration file. There are some legacy repositories that I work in that don't want CMake files to be formatted, it would be nice to add that in a .gersemirc in those repos so they won't be autoformatted without having to disable the tool globally.

clang-format has similar functionality.

**Comment by BlankSpruce** [comment]
I need you to elaborate further because perhaps I'm missing something. If you don't want to use formatter then why do you need an option to tell it to not format instead of just not using it? What does it mean for you "to disable tool globally"?

**Comment by nlebedenco** [comment]
@BlankSpruce  This can be useful to avoid accidental changes when you have a complex source tree for which certain subdirectories are known to contain files that should not be formatted for some reason (for example, 3rd party source code, git submodules etc). Mind that in many cases, gersemi might not be directly invoked by the user in a command-line interface but integrated in an IDE, or indirectly invoked by a build automation script or a pre-commit hook so the user may not have a direct way of separating which files should be formatted and which should not in every occasion. Not to mention the risk of accidental changes. And even when those changes can be reverted the user may not be able to do so until a long time has been wasted formatting files that should have never been touched from the start. With clang-format for example we can disable formatting for anything under a certain folder by creating a file that contains:
```
DisableFormat: true
SortIncludes: Never
```
And in addition since version 18 we can use .clang-format-ignore for more control, similarly to how .gitignore works.

With cmake-format this could be accomplished by creating a .cmake-format file containing:
```
format:
  disable: true
```

I would argue in favor of at least a way to disable formatting under a folder so we can ensure developers will never be format sensitive files by accident regardless of the integration used.

**Comment by BlankSpruce** [comment]
I'll think about the original request however it'd require to implement finding `.gersemirc` closest to the formatted file instead of the current approach. You might (or not) have noticed that configuration file is selected from the most common directory among formatted files. Example:

```shell
$ pwd
/tmp/my-project

$ gersemi foo/CMakeLists.txt
# candidate configuration files: 
# - /tmp/my-project/foo/.gersemirc
# - /tmp/my-project/.gersemirc
# - /tmp/.gersemirc
# - /.gersemirc

$ gersemi foo/CMakeLists.txt bar/CMakeLists.txt
# candidate configuration files: 
# - /tmp/my-project/.gersemirc
# - /tmp/.gersemirc
# - /.gersemirc

$ gersemi /tmp
# candidate configuration files:
# - /tmp/.gersemirc
# - /.gersemirc
```

I thought this is good enough approach because I don't think I'd like to have fragmented configuration (this directory formatted like such and such, that directory like such and such). Having mentioned all of that perhaps more suitable approach would be to introduce `exclude: <regex>|<regexes>`  to ignore certain files/directories.

I can see some merit of "disable formatter" now but I need to think what's the best to suit those needs.

**Comment by qweasdzxc787** [comment]
What I would suggest for processing the configuration settings: when you
are running against a file look at the command line parameters first for
overrides, then the current working directory of the file to be processed
for a config file, and then upwards every directory until you reach the
root directory to see if there’s a config file in the path of the current
working directory. If no config files are found use the default formatting
settings baked in to Graeme. That allows a for the use case of a monorepo
of many software projects to share a single gersemi configuration file, and
also allows the configuration to be overridden at every level. To make
processing simpler you should probably check the real path of the directory
so you don’t get into a softlink infinite loop in the file system
structure.

A debug mode that shows more information about what is being used for
formatting could show the path of the found nearest configuration file and
what settings are being used would be nice in general, doesn’t look like
that’s currently possible with gersemi.

Another nice addition to gersemi would be the ability to recursively
process all CMakeLists.txt files found in the current directory and
downwards (respecting the above override configuration structure) so you
could one shot format an entire projects CMake instead of having to
manually format each file. That also helps when a new gersemi formatting
option is released or a formatting change is made to a configuration file.

This tool has been very helpful for maintaining a consistent (beautiful!)
CMake style format in and across projects. Build systems are often an
afterthought or a necessary evil for most developers, but they are a
critical, foundational part of the code and deserve the same level of
tooling as the code itself. Really appreciate the work @BlankSpruce.

**Comment by BlankSpruce** [comment]
> I wasn’t aware that each directory could contain a .gersemirc file (documentation clarification needed). With that said, we’d still need a disable formatting flag in the file for the desired behavior.

Documentation will be enhanced on the next opportunity. I will make it clear that the most common configuration file is taken into consideration (in contrast to clang-format which takes the closest one if I recall correctly). Also in my previous answer I'm arguing that `exclude` will probably be better than just `disable` due to example provided by @nlebedenco and simply because it won't require me to change selection method of configuration file. I'd need stronger argument for change from "most common config file" to "closest config file".

> A debug mode that shows more information about what is being used for
formatting could show the path of the found nearest configuration file and
what settings are being used would be nice in general, doesn’t look like
that’s currently possible with gersemi.

I've been thinking about it for some time so this will happen in some form.

> Another nice addition to gersemi would be the ability to recursively
process all CMakeLists.txt files found in the current directory and
downwards (respecting the above override configuration structure) so you
could one shot format an entire projects CMake instead of having to
manually format each file. That also helps when a new gersemi formatting
option is released or a formatting change is made to a configuration file.

Unless I'm misunderstanding what you mean this has been the reality for quite some time already. In fact, due to startup overhead, it's always been preferable to pass all the files or directories in one invocation to the point I've made sure that there is `requires_serial: true`  in [pre-commit hook definition](https://github.com/BlankSpruce/gersemi/blob/master/.pre-commit-hooks.yaml#L6). Similarly the [usage description](https://github.com/BlankSpruce/gersemi?tab=readme-ov-file#usage) mentions that each positional argument is either file or directory. These are perfectly fine invocations:
```shell
# check all CMakeLists.txt and *.cmake files in this or child directory
$ gersemi -c . 

# check 
# - all CMakeLists.txt and *.cmake files in foo and bar directories (assuming these are directories)
# - ./CMakeLists.txt
# - baz/foo/utilities.cmake
$ gersemi -c foo CMakeLists.txt baz/foo/utilities.cmake bar

# check each file in separate invocation (less efficient but possible)
$ gersemi -c CMakeLists.txt
$ gersemi -c baz/foo/utilities.cmake
$ gersemi -c whatever.cmake
```
I'd recommend having only one `.gersemirc` per project, like at project's root, so that these invocations are equivalent in outcome.

**Comment by nlebedenco** [comment]
> I'd recommend having only one .gersemirc per project, like at project's root, so that these invocations are equivalent in outcome.

This cannot be controlled in projects with external dependencies. If I have a project A that depends on project B and it has its own .gersemirc I'll be locked out. For example:
```
Project A
  |__ .gersemirc
  |__  Project B
          |__ .gersemirc
  |__  Project C
    ...
```
Also consider cases where the files passed do not share a common root (which can happen on Windows) or only share the system root (/) which can happen on Unixes. This is very common in setups where CMake modules are stored in a separate repo shared by multiple other projects in the same company.

The common root approach is probably too opinionated and too simplistic. Note that even python Black suffers from it (aggravated by the fact that they decided to assume a .git folder also marks the project root regardless of the path passed in the --config argument, again so simplistic it makes it impossible even to define proper exclusion rules because Black internally matches the regex to relative paths based on what it thinks is the project root even when the files passed in the command-line have absolute paths).

**Comment by BlankSpruce** [comment]
Regarding closest configuration file approach there are a couple of things that aren't obvious to me how should I deal with them. Let's consider the following structure :
```
project
├── bar
│   ├── CMakeLists.txt
│   └── .gersemirc # [A]
├── CMakeLists.txt
├── foo
│   ├── CMakeLists.txt
│   └── something_else.txt
├── .gersemirc # [B]
└── something.cmake
```
Let's consider this invocation inside `project` directory:
```
$ gersemi --check .
```
1) What should be done if there is `workers: 4` in [A] config and `workers: 1` in [B] config? Should it lead to checking some set of files with 4 workers and some set of files with 1 worker?
2) What about `color` as in `$ gersemi --diff .`? Let's say there's `color: true` in [A] and `color: false` in [B]. Should the output be partially colored? Completely colored? Not colored at all?
3) Not as weird but similar things are to be considered about `unsafe` or `quiet`.

This indicates that there are two kinds of settings at play:
- those affecting application behaviour: `color`, `quiet`, `unsafe`, `workers`
- those affecting the outcome 
  - output: `definitions`, `indent`, `line_length`, `list_expansion`
  - useful side effects: `cache`, `warn_about_unknown_commands`

Perhaps I shouldn't conflate in one file so different things but also as a user I must say it's damn convenient this way. Perhaps behaviour affecting the settings should only follow `default < command line` chain and those affecting outcome `default < configuration file < command line`. I'm not comfortable enough right now in breaking that already established modus operandi even if it has some other shortcomings.

Regarding `disable`: I'm slowly convincing myself that `disable` is superior approach than `exclude` because proper selection of files can always be implemented by user in a script in any way they see fit. For example:
`$ gersemi CMakeLists.txt *.cmake foo` will perfectly deal with excluding `bar` directory without explicitly listing all the files in `foo` directory. At the same time `disable` would be rather convenient for single file usage like it happens in editors and IDEs. However the big caveat is that without changing approach from "most common configuration file" to "closest configuration file" it's not going to be useful in practice. I've backed myself into the corner and I don't know right now how to make it nice but not disruptive nor surprising to current users.

**Comment by petk** [comment]
Hello, as this topic is opened. Would it be also possible to add a command-line configuration option to select specific configuration .gersemirc file from a given location?

Perhaps something like:

```sh
gersemi --config path/to/.gersemirc src ...
```

**Comment by BlankSpruce** [comment]
@petk Once I decide on how to proceed with feature requested in this issue I'll definitely consider `--config`. I don't want to take care of extra work before that happens and as you can see there is no clear winner yet. I need some time to ponder but it's a bit difficult right now for me.

**Comment by BlankSpruce** [comment]
On [next-release](https://github.com/BlankSpruce/gersemi/tree/next-release) branch there is new version of gersemi with the following changes:
- configuration file is now selected by searching for the closest file, similarly how it's done in `rustfmt` or `ruff`
- `cache`, `color`, `quiet` and `workers` are degraded to command line only options because, in the hindsight, it doesn't make that much sense to share them with others and it saves me from attempting to find reasonable interpretation of those when multiple configuration files are at play
- given that some configuration files in the wild do contain the degraded options I've added soft warnings about their presence, their values in the configuration file won't be taken into consideration
- `--config` is available, it'll take precedence over finding relevant `.gersemirc` file
- `disable_formatting` is available
- configuration part of `--help` is divided into two section, "outcome configuration" and "control configuration"

I haven't done `--print-config` and improvements to README yet but I intend to finish that before I release next version. I'm going to postpone that release to have some time for extra testing. Any information about bugs related to aforementioned features or unclear description/improper wording in `--help`  will be appreciated.

**Comment by BlankSpruce** [comment]
All the things in my previous comment [and a couple more](https://github.com/BlankSpruce/gersemi/blob/master/CHANGELOG.md#0170-2024-10-26) are available in 0.17.0. If there is any problem with these new features please open new issue instead of commenting here.

**Comment by petk** [comment]
Thank you so much @BlankSpruce according to some quick testing it works very nicely
